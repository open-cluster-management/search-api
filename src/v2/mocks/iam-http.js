/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */
/* eslint-disable global-require */

export default function createMockHttp() {
  const state = {
    namespaces: {
      kind: 'ProjectList',
      apiVersion: 'project.openshift.io/v1',
      metadata: { selfLink: '/apis/project.openshift.io/v1/projects' },
      items: [
        {
          metadata: {
            name: 'default',
            selfLink: '/apis/project.openshift.io/v1/projects/default',
            uid: 'd039f9ea-152f-11ea-920f-00000a1012a4',
            resourceVersion: '21120909',
            creationTimestamp: '2019-12-02T18:16:08Z',
          },
          spec: { finalizers: [] },
          status: { phase: 'Active' },
        },
        {
          metadata: {
            name: 'kube-system',
            selfLink: '/apis/project.openshift.io/v1/projects/kube-system',
            uid: 'ce32a2c7-152f-11ea-920f-00000a1012a4',
            resourceVersion: '21133087',
            creationTimestamp: '2019-12-02T18:16:05Z',
          },
          spec: { finalizers: [] },
          status: { phase: 'Active' },
        },
      ],
    },
    username: {
      body: {
        kind: 'TokenReview',
        apiVersion: 'authentication.k8s.io/v1',
        metadata: { creationTimestamp: null },
        spec: {
          token: 'testToken',
        },
        status: {
          authenticated: true,
          user: {
            username: 'kube:admin',
          },
        },
      },
      attempts: 1,
    },
  };

  return async function MockLib(params) {
    switch (true) {
      case params.url.includes('project.openshift.io/v1/projects'):
        return state.namespaces;
      case params.url.includes('/apis/authentication.k8s.io/v1/tokenreviews'):
        return state.username;
      default:
        return state.namespaces;
    }
  };
}

