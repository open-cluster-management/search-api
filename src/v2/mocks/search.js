/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */
// Copyright Contributors to the Open Cluster Management project
// eslint-disable-next-line max-classes-per-file
import _ from 'lodash';
import RedisGraphConnector from '../connectors/redisGraph';

const resourceTemplate = {
  cluster: (id) => ({
    kind: 'cluster',
    name: `mock-cluster${id > 0 ? `-${id}` : ''}`,
    namespace: 'mock-mcm-cluster',
    status: 'ok',
    endpoint: '',
    nodes: 10,
    cpu: '30%',
    memory: '20%',
    storage: '10%',
  }),
  node: (id = 0) => ({
    kind: 'node',
    name: `mock-1.1.1.${id}`,
    cpus: 10,
    role: ['master'],
  }),
  pod: (id) => ({
    kind: 'pod',
    name: `mock-cluster${id > 0 ? `-${id}` : ''}`,
    namespace: 'kube-system',
    restarts: 20,
    creationTimestamp: '2018-09-27 16:52:20 -0400 EDT',
    status: 'Running',
    hostIP: '179.160.35.59',
    podIP: '179.160.35.59',
  }),
};

// export function repeat(map, count) {
//   return new Array(count).fill(0).map(map);
// }

export const mockSearchResult = {
  /**
   * Mock search results.
   *
   * Pass the kind and amount of objects to mock.
   * For example, to mock a search that returns 2 clusters and 5 pods:
   *    { cluster: 2, pod:5 }
   */
  mock: (args) => {
    const mockResult = [];

    // Get items
    Object.keys(args).forEach((kind) => {
      for (let i = 0; i < args[kind]; i += 1) {
        const mockItem = { ...resourceTemplate[kind](i) };
        mockResult.push(mockItem);
      }
    });

    return mockResult;
  },
};

const MOCK_QUERIES = {
  /*
   * Application queries mocks.
   */
  runApplicationsQuery: [
    {
      'app._uid': 'local-cluster/29a848d6-3de8-11ea-9f0f-00000a100f99',
      'app.created': '2020-01-23T13:56:32Z',
      'app.dashboard':
        'localhost/grafana/dashboard/db/app01-dashboard-via-federated-prometheus?namespace=sample',
      'app.label': 'label1=value1; label2=value2',
      'app.name': 'app01',
      'app.namespace': 'sample',
      'app.selfLink':
        '/apis/app.k8s.io/v1beta1/namespaces/sample/applications/app01',
    },
    {
      'app._uid': 'local-cluster/app-02-uid',
      'app.created': '2020-01-23T13:56:32Z',
      'app.dashboard':
        'localhost/grafana/dashboard/db/app02-dashboard-via-federated-prometheus?namespace=test',
      'app.name': 'app02',
      'app.namespace': 'test',
      'app.selfLink':
        '/apis/app.k8s.io/v1beta1/namespaces/test/applications/app02',
    },
  ],

  runAppClustersQuery: [
    {
      'app._uid': 'local-cluster/29a848d6-3de8-11ea-9f0f-00000a100f99',
      local: false,
      clusterCount: 2,
    },
    {
      'app._uid': 'local-cluster/29a848d6-3de8-11ea-9f0f-00000a100f99',
      local: true,
      clusterCount: 1,
    },
  ],

  runAppHubSubscriptionsQuery: [
    {
      'app._uid': 'local-cluster/29a848d6-3de8-11ea-9f0f-00000a100f99',
      'sub._uid': 'local-cluster/bdced01f-3bd4-11ea-a488-00000a100f99',
      'sub.name': 'sub1',
      'sub.namespace': 'default',
      'sub.channel': 'dev1/dev1',
      'sub.localPlacement': 'true',
    },
    {
      'app._uid': 'local-cluster/29a848d6-3de8-11ea-9f0f-00000a100f99',
      'sub._uid': 'local-cluster/b218636d-3d5e-11ea-8ed1-00000a100f99',
      'sub.name': 'sub2',
      'sub.namespace': 'default',
      'sub.status': 'Propagated',
      'sub.channel': 'default/mortgage-channel',
      'sub.timeWindow': 'blocked',
    },
    {
      'app._uid': 'local-cluster/29a848d6-3de8-11ea-9f0f-00000a100f99',
      'sub._uid': 'local-cluster/66426f24-3bd3-11ea-a488-00000a100f99',
      'sub.name': 'sub3',
      'sub.namespace': 'default',
      'sub.status': 'Propagated',
      'sub.channel': 'dev1/dev1',
      'sub.timeWindow': 'active',
    },
  ],

  runAppHubChannelsQuery: [
    {
      'app._uid': 'local-cluster/29a848d6-3de8-11ea-9f0f-00000a100f99',
      'sub._uid': 'local-cluster/bdced01f-3bd4-11ea-a488-00000a100f99',
      'sub.name': 'sub1',
      'sub.namespace': 'default',
      'sub._gitbranch': 'master',
      'sub._gitpath': 'helloworld',
      'sub._gitcommit': 'd67d8e10dcfa41dddcac14952e9872e1dfece06f',
      'ch._uid': 'local-cluster/233dfffd-f421-44ee-811b-7f3352b2d728',
      'ch.type': 'Git',
      'ch.pathname': 'https://github.com/fxiang1/app-samples.git',
    },
    {
      'app._uid': 'local-cluster/29a848d6-3de8-11ea-9f0f-00000a100f99',
      'sub._uid': 'local-cluster/b218636d-3d5e-11ea-8ed1-00000a100f99',
      'sub.name': 'sub2',
      'sub.namespace': 'default',
      'ch._uid': 'local-cluster/6c8dcb97-5e6e-4606-9b55-ae3eb05fcfb5',
      'ch.type': 'Namespace',
      'ch.pathname': 'sample-ns',
    },
    {
      'app._uid': 'local-cluster/app-02-uid',
      'sub._uid': 'local-cluster/66426f24-3bd3-11ea-a488-00000a100f99',
      'sub.name': 'sub3',
      'sub.namespace': 'default',
      'sub.package': 'application-chart',
      'sub.packageFilterVersion': '2.2.0',
      'ch._uid': 'local-cluster/233dfffd-f421-44ee-811b-7f3352b2d728',
      'ch.type': 'HelmRepo',
      'ch.pathname':
        'http://multiclusterhub-repo.open-cluster-management.svc.cluster.local:3000/charts',
    },
  ],

  /*
   * Subscription queries mocks.
   */
  runSubscriptionsQuery: [
    {
      'sub._uid': 'local-cluster/sub-01-uid',
      'sub.name': 'sub01',
      'sub.namespace': 'applications',
      'sub.created': '2020-08-20T14:16:05Z',
      'sub.selfLink': '/apis/apps.open-cluster-management.io/v1/namespaces/applications/sub01',
      'sub.status': 'Propagated',
      'sub.channel': 'git-ch-ns/git-ch',
      'sub.timeWindow': 'blocked',
      'sub.localPlacement': 'true',
    },
    {
      'sub._uid': 'local-cluster/sub-02-uid',
      'sub.name': 'sub02',
      'sub.namespace': 'test',
      'sub.created': '2020-08-20T14:17:05Z',
      'sub.selfLink': '/apis/apps.open-cluster-management.io/v1/namespaces/test/sub02',
      'sub.status': 'PropagationFailed',
      'sub.channel': 'object-ch-ns/object-ch',
      'sub.localPlacement': 'false',
    },
  ],

  runSubClustersQuery: [
    {
      'sub._uid': 'local-cluster/sub-01-uid',
      local: false,
      clusterCount: 2,
    },
    {
      'sub._uid': 'local-cluster/sub-01-uid',
      local: true,
      clusterCount: 1,
    },
    {
      'sub._uid': 'local-cluster/sub-02-uid',
      local: false,
      clusterCount: 2,
    },
  ],

  runSubAppsQuery: [
    {
      'sub._uid': 'local-cluster/sub-01-uid',
      count: 1,
    },
    {
      'sub._uid': 'local-cluster/sub-02-uid',
      count: 2,
    },
  ],

  /*
   * Placement rules queries mocks.
   */
  runPlacementRulesQuery: [
    {
      'pr._uid': 'local-cluster/pr-01-uid',
      'pr.name': 'pr01',
      'pr.namespace': 'applications',
      'pr.created': '2020-08-20T14:16:05Z',
      'pr.selfLink': '/apis/apps.open-cluster-management.io/v1/namespaces/placementrules/pr01',
      'pr.replicas': 3,
    },
    {
      'pr._uid': 'local-cluster/pr-02-uid',
      'pr.name': 'pr02',
      'pr.namespace': 'test',
      'pr.created': '2020-08-20T14:17:05Z',
      'pr.selfLink': '/apis/apps.open-cluster-management.io/v1/namespaces/test/pr02',
    },
  ],

  runPRClustersQuery: [
    {
      'pr._uid': 'local-cluster/sub-01-uid',
      local: false,
      clusterCount: 2,
    },
    {
      'pr._uid': 'local-cluster/sub-01-uid',
      local: true,
      clusterCount: 1,
    },
    {
      'pr._uid': 'local-cluster/sub-02-uid',
      local: false,
      clusterCount: 2,
    },
  ],

  /*
   * Channels queries mocks.
   */
  runChannelsQuery: [
    {
      'ch._uid': 'local-cluster/ch-01-uid',
      'ch.name': 'ch01',
      'ch.namespace': 'applications',
      'ch.created': '2020-08-20T14:16:05Z',
      'ch.selfLink': '/apis/apps.open-cluster-management.io/v1/namespaces/applications/channels/ch01',
      'ch.type': 'HelmRepo',
      'ch.pathname': 'http://multiclusterhub-repo.open-cluster-management.svc.cluster.local:3000/charts',
    },
    {
      'ch._uid': 'local-cluster/ch-02-uid',
      'ch.name': 'ch02',
      'ch.namespace': 'test',
      'ch.created': '2020-08-20T14:17:05Z',
      'ch.selfLink': '/apis/apps.open-cluster-management.io/v1/namespaces/test/channels/ch02',
      'ch.type': 'Git',
      'ch.pathname': 'https://github.com/fxiang1/app-samples.git',
    },
  ],

  runChannelSubsQuery: [
    {
      'ch._uid': 'local-cluster/ch-01-uid',
      localPlacement: ['true', 'false'],
      count: 0,
    },
    {
      'ch._uid': 'local-cluster/ch-02-uid',
      localPlacement: ['false'],
      count: 3,
    },
  ],

  runChannelClustersQuery: [
    {
      'ch._uid': 'local-cluster/ch-01-uid',
      local: false,
      clusterCount: 2,
    },
    {
      'ch._uid': 'local-cluster/ch-01-uid',
      local: true,
      clusterCount: 1,
    },
    {
      'ch._uid': 'local-cluster/ch-02-uid',
      local: false,
      clusterCount: 2,
    },
  ],
};

export default class MockSearchConnector extends RedisGraphConnector {
  constructor(args) {
    super(args);

    this.g = {
      query: (queryString) => {
        throw new Error(`No mock data for query: ${queryString}`);
      },
    };

    // Set up query functions that use the actual query-building code in RedisGraphConnector
    // but return the result from static data in this file
    Object.entries(MOCK_QUERIES).forEach(([key, value]) => {
      this[key] = () => {
        /* eslint-disable class-methods-use-this */
        const query = new class extends MockSearchConnector {
          constructor() {
            super(args);
            this.g = {
              query: () => {
                const clonedValue = _.cloneDeep(value);
                const iterator = {
                  hasNext: () => clonedValue.length > 0,
                  next: () => {
                    const record = new Map();
                    Object.entries(clonedValue.shift()).forEach(([mapKey, mapValue]) => {
                      record.set(mapKey, mapValue);
                    });
                    return {
                      keys: () => Array.from(record.keys()),
                      get: (k) => record.get(k),
                    };
                  },
                };
                return iterator;
              },
            };
          }
        }();
        return RedisGraphConnector.prototype[key].call(query);
      };
    });
  }

  async isServiceAvailable() {
    return true;
  }

  // Search Query Mocks
  async runSearchQuery() {
    return mockSearchResult.mock({ cluster: 2, node: 3, pod: 5 });
  }

  async runSearchQueryCountOnly() {
    return 10;
  }

  async getAllProperties() {
    return ['kind', 'name', 'namespace', 'cpu', 'created'];
  }

  async getAllValues(property) {
    switch (property) {
      case 'kind':
        return ['cluster', 'application', 'deployable', 'compliance'];
      case 'namespace':
        return ['default', 'kube-system'];
      case 'cpu':
        return ['<', '>', '!='];
      case 'created':
        return ['hour', 'day', 'week', 'month', 'year'];
      default:
        return ['kind', 'name', 'namespace', 'cpu', 'created'];
    }
  }

  async findRelationships() {
    return [{
      kind: 'cluster',
      count: 1,
      items: [{
        _uid: '123', kind: 'cluster', name: 'mockCluster', namespace: 'mockCluster-ns',
      }],
    }];
  }

  async getRbacValues() {
    return { allowedResources: [], allowedNS: [] };
  }
}
