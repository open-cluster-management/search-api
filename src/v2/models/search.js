/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project

import _ from 'lodash';
import config from '../../../config';
import { isRequired } from '../lib/utils';
import logger from '../lib/logger';
import getOperatorStatus from './operatorStatus';

// Remove single and double quotes because these can be used to inject malicious
// code in the RedisGraph query. (SQL injection).
function sanitizeString(s) {
  return s.replace(/['"]/g, '');
}

// Sanitize all inputs to prevent "sql injection" attacks.
function sanitizeInputs({
  keywords = [],
  filters = [],
  property = '',
  limit,
  relatedKinds = [],
}) {
  const sanitizedKeywords = keywords.map((k) => sanitizeString(k));
  const sanitizedFilters = filters.map((f) => ({
    property: sanitizeString(f.property),
    values: f.values.map((v) => sanitizeString(v)),
  }));

  return {
    keywords: sanitizedKeywords,
    filters: sanitizedFilters,
    property: sanitizeString(property),
    limit: limit || config.get('defaultQueryLimit'),
    relatedKinds: relatedKinds.map((k) => sanitizeString(k)),
  };
}

// TODO: Keyword filtering currently requires that we transfer a large number of records from
// RedisGraph to filter locally. We need to investigate alternatives to improve performance.
function filterByKeywords(resultSet, keywords) {
  /* Regular expression resolves to a string like:
   *     /(?=.*keyword1)(?=.*keyword2)(?=.*keyword3)/gi
   * which matches if the string contains all keywords and is case insensitive. */
  const regex = new RegExp(keywords.reduce((prev, curr) => `${prev}(?=.*${curr})`, ''), 'gi');

  return resultSet.filter((r) => Object.values(r).toString().match(regex));
}

export default class SearchModel {
  constructor({ searchConnector = isRequired('searchConnector'), kubeConnector = isRequired('kubeConnector') }) {
    this.kubeConnector = kubeConnector;
    this.searchConnector = searchConnector;
  }

  async checkSearchServiceAvailable() {
    const isServiceAvailable = await this.searchConnector.isServiceAvailable();
    if (!isServiceAvailable) {
      const deployRedisgraph = await getOperatorStatus(this.kubeConnector);
      if (!deployRedisgraph) {
        logger.warn('The search service is not enabled in the current configuration.');
        throw Error('The search service is not enabled in the current configuration.');
      } else {
        logger.error('Unable to resolve search request because RedisGraph is unavailable.');
        throw Error('Search service is unavailable.');
      }
    }
  }

  async searchQueryLimiter(keywords, filters, limit) {
    let exitLoop = false;
    let querySkipIdx = 0;
    let results = [];
    while (!exitLoop) {
      // eslint-disable-next-line
      const searchResults = await this.searchConnector.runSearchQuery(filters, limit, querySkipIdx);
      // Filter results if keyword search - otherwise its a label search
      const filteredResults = keywords.length > 0
        ? filterByKeywords(searchResults, keywords)
        : searchResults;
      // concatenate search results each iteration removing any duplicates
      results = _.unionBy(results, filteredResults, '_uid');
      querySkipIdx += 1;
      if (searchResults.length < config.get('defaultQueryLoopLimit') || results.length >= limit) exitLoop = true;
    }
    return results;
  }

  async resolveSearch(input) {
    const {
      keywords,
      filters,
      limit,
    } = sanitizeInputs(input);
    await this.checkSearchServiceAvailable();
    if (keywords && keywords.length > 0) {
      return this.searchQueryLimiter(keywords, filters, limit);
    }
    return this.searchConnector.runSearchQuery(filters, limit, -1);
  }

  async resolveSearchCount(input) {
    const {
      keywords,
      filters,
    } = sanitizeInputs(input);
    await this.checkSearchServiceAvailable();
    if (keywords && keywords.length > 0) {
      const results = await this.searchQueryLimiter(keywords, filters, -1);
      return results.length;
    }
    return this.searchConnector.runSearchQueryCountOnly(filters);
  }

  async resolveSearchComplete(input, limit) {
    const { property, filters } = sanitizeInputs(input);
    await this.checkSearchServiceAvailable();
    return this.searchConnector.getAllValues(property, filters, limit);
  }

  /** Resolve the related items for a given search query.
   *
   * @param {*} parent
   * returns { kind: String, count: Int, items: [] }
   */
  async resolveRelated(input, countOnly) {
    const { filters, relatedKinds } = sanitizeInputs(input);
    await this.checkSearchServiceAvailable();
    const relationships = await this.searchConnector.findRelationships({ filters, countOnly, relatedKinds });

    const result = {};
    relationships.forEach((r) => {
      if (!result[r.kind]) {
        result[r.kind] = [];
      }
      result[r.kind].push(r);
    });

    const resultKinds = Object.keys(result);

    // TODO: Need a better prioritization algorithm.
    const prioritizedRelationships = [
      'application',
      'cluster',
      'release',
      'node',
      'persistentvolume',
      'pod',
      'secret',
      'persistentvolumeclaim',
    ];
    resultKinds.sort((a, b) => prioritizedRelationships.indexOf(b));

    return resultKinds.map((r) => ({ kind: r, count: result[r].length, items: result[r] }));
  }

  async searchSchema() {
    await this.checkSearchServiceAvailable();
    return {
      allProperties: await this.searchConnector.getAllProperties(),
    };
  }
}
