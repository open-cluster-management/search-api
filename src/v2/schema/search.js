/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 * Copyright (c) 2020 Red Hat, Inc.
 ****************************************************************************** */
// Copyright Contributors to the Open Cluster Management project
import lodash from 'lodash';
import { gql } from 'apollo-server-express';
import logger from '../lib/logger';

export const typeDef = gql`
  input SearchFilter {
    property: String!
    values: [String]
  }

  input SearchInput {
    keywords: [String]
    filters: [SearchFilter]
    # Max number of results. Default limit: 10,000. For unlimited results use -1.
    limit: Int
    # Filter relationships to the specified kinds.
    # If empty, all relationships will be included. This filter is used with the 'related' field on SearchResult.
    relatedKinds: [String]
  }

  type SearchResult {
    count: Int
    items: JSON
    related: [SearchRelatedResult]
  }

  type SearchRelatedResult {
    kind: String!
    count: Int
    items: JSON
  }
`;

export const resolver = {
  Query: {
    search: (parent, { input }) => input,
    searchComplete: (parent, { property, query = {} }, { searchModel }) => searchModel.resolveSearchComplete({
      property,
      filters: lodash.get(query, 'filters', []),
    }, lodash.get(query, 'limit', 0)),
    searchSchema: (parent, args, { searchModel }) => searchModel.searchSchema(),
  },
  SearchResult: {
    count: (parent, args, { searchModel }) => searchModel.resolveSearchCount(parent),
    items: (parent, args, { searchModel }) => searchModel.resolveSearch(parent),
    related: (parent, args, { searchModel }, info) => {
      const selections = lodash.get(info, 'fieldNodes[0].selectionSet.selections', [])
        .map((s) => lodash.get(s, 'name.value', []));
      const countOnly = selections.includes('count') && !selections.includes('items');
      if (selections.includes('count') && selections.includes('items')) {
        // eslint-disable-next-line max-len
        logger.warn('Client requested related items and count in the same query. When both are needed clients should get the count from items.length for better performance.');
      }
      return searchModel.resolveRelated(parent, countOnly);
    },
  },
};
