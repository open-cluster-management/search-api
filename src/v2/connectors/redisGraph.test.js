/* eslint-disable no-underscore-dangle */
/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

// import mockRedis from 'redis-mock';
import { Record } from 'redisgraph.js';
import ResultSet from 'redisgraph.js/src/resultSet';
import { mockSearchResult } from '../mocks/search';
import RedisGraphConnector, {
  formatResult, getOperator, getDateFilter, getFilterString, isDate,
} from './redisGraph';

describe('redisGraph', () => {
  describe('Class Functions', () => {
    test('getOperator', async () => {
      expect(getOperator('<=')).toEqual('<=');
      expect(getOperator('>=')).toEqual('>=');
      expect(getOperator('<')).toEqual('<');
      expect(getOperator('>')).toEqual('>');
      expect(getOperator('!')).toEqual('<>');
      expect(getOperator('!=')).toEqual('<>');
    });
    test('getFilterString', async () => {
      Date.now = jest.fn(() => 1548076708000);
      expect(getFilterString([{ property: 'kind', values: ['cluster'] }])).toEqual('(n.kind = \'cluster\')');
      expect(getFilterString([{ property: 'role', values: ['master'] }])).toEqual('((\'master\' IN n.role))');
      expect(getFilterString([{ property: 'cpu', values: ['<16'] }])).toEqual('(n.cpu < 16)');
      expect(getFilterString([{ property: 'created', values: ['month'] }])).toEqual('(n.created > \'2018-12-22T02:49:25.000Z\')');
    });
    test('getDateFilter', async () => {
      Date.now = jest.fn(() => 1548076708000); // 21-01-2019T13:18:28Z
      expect(getDateFilter('hour')).toEqual('> \'2019-01-21T12:18:28.000Z\'');
      expect(getDateFilter('day')).toEqual('> \'2019-01-20T13:18:28.000Z\'');
      expect(getDateFilter('week')).toEqual('> \'2019-01-14T13:18:28.000Z\'');
      expect(getDateFilter('month')).toEqual('> \'2018-12-22T02:49:25.000Z\'');
      expect(getDateFilter('year')).toEqual('> \'2018-01-21T07:29:42.000Z\'');
      expect(getDateFilter('default')).toEqual('> \'2018-12-22T02:49:25.000Z\'');
    });
    test('isDate', async () => {
      expect(isDate('2016-06-23T09:07:21-07:00')).toEqual(true);
    });
    test('formatResult', async () => {
      const results = new ResultSet();

      results._results = [new Record([' n '], [mockSearchResult.mock({ node: 5 })])];
      results._resultsCount = results._results.length;

      results.next().get(' n ').properties = [{
        kind: 'node',
        role: ['master'],
        name: 'mock-1.1.1.0',
        cpus: 10,
      }];

      results._position = 0;
      expect(formatResult(results, true)).toMatchSnapshot();

      results._position = 0;
      expect(formatResult(results, false)).toMatchSnapshot();
    });
  });

  /**
   * createWhereClause
   *    - getRbacString
   *      - checkIfOpenShiftPlatform
   *      - getUserAccess
   *      - getNonNamespacedAccess
   *        - getNonNamespacedResources
   * runSearchQuery
   * runSearchQueryCountOnly
   * getAllProperties
   * getAllValues
   * findRelationships
   */

  // eslint-ignore jest/no-commented-out-tests
  describe('Testing Connector Functions', () => {
    const searchConnector = new RedisGraphConnector({ rbac: ['kube-system', 'default'], req: { user: { name: 'TestUserName' }, kubeToken: 'Bearer localdev' } });
    const _ = new Promise((resolve) => resolve({}));

    // test('isServiceAvailable', () => searchConnector.isServiceAvailable().then((res) => expect(res).toBe(true)));
    // test('getAllProperties', async () => {
     // const values = ['cluster', 'kind', 'label', 'name', 'namespace', 'status'];
     // const properties = await searchConnector.getAllProperties();

     // values.forEach((val) => {
       // expect(properties.includes(val)).toBe(true);
     // });
   // });

    test('getAllValues', async () => {
      expect(searchConnector.getAllValues('', [])).toEqual(_);
      // With mock data, we won't be able to test it without the proper (withclause and whereclause)
      expect(searchConnector.getAllValues('role', [{ property: 'role', values: ['master'] }])).toEqual(_);
    });

    // test('Check If Redis Is Available', async () => {
    //   expect(await searchConnector.isServiceAvailable()).toBe(true);
    // });

  //   test('Create the RBAC String for Redis Queries', async () => {
  //     const RBACString =
  //        await searchConnector.createWhereClause([{ property: 'pods', values: ['testPod'] }]);
  //     console.log(RBACString);
  //   });
  });
});
