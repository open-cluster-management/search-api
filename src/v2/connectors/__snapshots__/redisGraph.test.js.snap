// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redisGraph Class Functions formatResult 1`] = `
Array [
  Array [
    Object {
      "cpus": 10,
      "kind": "node",
      "name": "mock-1.1.1.0",
      "role": Array [
        "master",
      ],
    },
  ],
]
`;

exports[`redisGraph Class Functions formatResult 2`] = `
Array [
  Object {
    " n ": Array [
      Object {
        "cpus": 10,
        "kind": "node",
        "name": "mock-1.1.1.0",
        "role": Array [
          "master",
        ],
      },
      Object {
        "cpus": 10,
        "kind": "node",
        "name": "mock-1.1.1.1",
        "role": Array [
          "master",
        ],
      },
      Object {
        "cpus": 10,
        "kind": "node",
        "name": "mock-1.1.1.2",
        "role": Array [
          "master",
        ],
      },
      Object {
        "cpus": 10,
        "kind": "node",
        "name": "mock-1.1.1.3",
        "role": Array [
          "master",
        ],
      },
      Object {
        "cpus": 10,
        "kind": "node",
        "name": "mock-1.1.1.4",
        "role": Array [
          "master",
        ],
      },
    ],
  },
]
`;
